pkg/chunk/grpc/storage_client.go

```
@@ -0,0 +1,128 @@
+	for i := range chunks {
+		buf, err := chunks[i].Encoded()
+		if err != nil {
+			return errors.WithStack(err)
+		}
+
+		key := chunks[i].ExternalKey()
+		tableName, err := s.schemaCfg.ChunkTableFor(chunks[i].From)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		writeChunk := &Chunk{
+			Encoded:   buf,
+			Key:       key,
+			TableName: tableName,
+		}
+
+		chunksInfo.Chunks = append(chunksInfo.Chunks, writeChunk)
+	}
+
+	_, err := s.client.PutChunks(ctx, chunksInfo)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+func (s *StorageClient) DeleteChunk(ctx context.Context, chunkID string) error {
+	chunkInfo := &ChunkID{ChunkID: chunkID}
+	_, err := s.client.DeleteChunks(ctx, chunkInfo)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return nil
+}
+
+func (s *StorageClient) GetChunks(ctx context.Context, input []chunk.Chunk) ([]chunk.Chunk, error) {
+	chunks := &GetChunksRequest{}
+	chunks.Chunks = []*Chunk{}
+	var err error
+	for _, inputInfo := range input {
+		chunkInfo := &Chunk{}
+		// send the table name from upstream gRPC client as gRPC server is unaware of schema
+		chunkInfo.TableName, err = s.schemaCfg.ChunkTableFor(inputInfo.From)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		chunkInfo.Key = inputInfo.ExternalKey()
+		chunks.Chunks = append(chunks.Chunks, chunkInfo)
+	}
+	streamer, err := s.client.GetChunks(ctx, chunks)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	var result []chunk.Chunk
+	var decodeContext *chunk.DecodeContext
+	for {
+		receivedChunks, err := streamer.Recv()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		for _, chunkResponse := range receivedChunks.GetChunks() {
+			var c chunk.Chunk
+			if chunkResponse != nil {
+				decodeContext = chunk.NewDecodeContext()
+				err = c.Decode(decodeContext, chunkResponse.Encoded)
+				if err != nil {
+					return result, err
+				}
+			}
+			if c.Data != nil {
```

I think we can remove this check here.

---

@gouthamve for some reason if the gRPC server sends `c.Data `stream as nil cortex will crash.
So it's an additional check to ensure we append data which is not nil. If it's okay I can remove it.

---

I would remove it as well. Silently ignoring this condition may lead to very subdle bugs. Having Cortex panicking sounds actually better in this edge case.